File: openc3/lib/openc3/packets/parsers/packet_item_parser.rb
# encoding: ascii-8bit

# Copyright 2022 Ball Aerospace & Technologies Corp.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# Modified by OpenC3, Inc.
# All changes Copyright 2025, OpenC3, Inc.
# All Rights Reserved
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

require 'openc3/packets/packet_item'

module OpenC3
  # Parses a packet item definition and creates a new PacketItem
  class PacketItemParser
    # This number is a little arbitrary but there are definitely issues at
    # 1 million and you really shouldn't be doing anything this big anyway
    BIG_ARRAY_SIZE = 100_000

    # @param parser [ConfigParser] Configuration parser
    # @param packet [Packet] The packet the item should be added to
    # @param cmd_or_tlm [String] Whether this is a command or telemetry packet
    # @param warnings [Array<String>] Array of warning strings from PacketConfig
    def self.parse(parser, packet_config, packet, cmd_or_tlm, warnings)
      parser = PacketItemParser.new(parser, packet_config, warnings)
      parser.verify_parameters(cmd_or_tlm)
      parser.create_packet_item(packet, cmd_or_tlm)
    end

    # @param parser [ConfigParser] Configuration parser
    # @param warnings [Array<String>] Array of warning strings from PacketConfig
    def initialize(parser, packet_config, warnings)
      @parser = parser
      @packet_config = packet_config
      @warnings = warnings
      @usage = get_usage()
    end

    def verify_parameters(cmd_or_tlm)
      if @parser.keyword.include?('ITEM') && cmd_or_tlm == PacketConfig::COMMAND
        raise @parser.error("ITEM types are only valid with TELEMETRY", @usage)
      elsif @parser.keyword.include?('PARAMETER') && cmd_or_tlm == PacketConfig::TELEMETRY
        raise @parser.error("PARAMETER types are only valid with COMMAND", @usage)
      end

      # The usage is formatted with brackets <XXX> around each option so
      # count the number of open brackets to determine the number of options
      max_options = @usage.count("<")
      if @parser.keyword.include?('STRUCTURE')
        @parser.verify_num_parameters(max_options, max_options, @usage)
      else
        # The last two options (description and endianness) are optional
        @parser.verify_num_parameters(max_options - 2, max_options, @usage)
      end
      @parser.verify_parameter_naming(1) # Item name is the 1st parameter
    end

    def create_packet_item(packet, cmd_or_tlm)
      item_name = @parser.parameters[0].upcase
      if packet.items[item_name]
        msg = "#{packet.target_name} #{packet.packet_name} #{item_name} redefined."
        Logger.instance.warn msg
        @warnings << msg
      end
      if @parser.keyword.include?("STRUCTURE")
        item = PacketItem.new(item_name,
                              get_bit_offset(),
                              get_bit_size(true),
                              :BLOCK,
                              :BIG_ENDIAN,
                              nil,
                              :ERROR) # overflow
      else
        item = PacketItem.new(item_name,
                              get_bit_offset(),
                              get_bit_size(),
                              get_data_type(),
                              get_endianness(packet),
                              get_array_size(),
                              :ERROR) # overflow
        if cmd_or_tlm == PacketConfig::COMMAND
          item.range = get_range()
          item.default = get_default()
        end
        item.id_value = get_id_value(item)
        item.description = get_description()
      end
      if append?
        item = packet.append(item)
      else
        item = packet.define(item)
      end
      if @parser.keyword.include?("STRUCTURE")
        structure = lookup_packet(get_cmd_or_tlm(), get_target_name(), get_packet_name())
        packet.structurize_item(item, structure)
      end
      item
    end

    private

    def append?
      @parser.keyword.include?("APPEND")
    end

    def get_data_type
      index = append? ? 2 : 3
      @parser.parameters[index].upcase.to_sym
    end

    def get_bit_offset
      return 0 if append?
      Integer(@parser.parameters[1])
    rescue => e
      raise @parser.error(e, @usage)
    end

    def get_bit_size(check_structure = false)
      index = append? ? 1 : 2
      bit_size = @parser.parameters[index]
      if not check_structure or bit_size.to_s.upcase != 'DEFINED'
        return Integer(bit_size)
      else
        structure = lookup_packet(get_cmd_or_tlm(), get_target_name(), get_packet_name())
        return structure.defined_length_bits
      end
    rescue => e
      raise @parser.error(e, @usage)
    end

    def get_array_size
      return nil unless @parser.keyword.include?('ARRAY')

      index = append? ? 3 : 4
      array_bit_size = Integer(@parser.parameters[index])
      items = array_bit_size / get_bit_size()
      if items >= BIG_ARRAY_SIZE
        warning = "Performance Issue!\n"\
                  "In #{@parser.filename}:#{@parser.line_number} your definition of:\n"\
                  "#{@parser.line}\n"\
                  "creates an array with #{items} elements. Consider creating a BLOCK if this is binary data."
        Logger.warn(warning)
        @warnings << warning
      end
      array_bit_size
    rescue => e
      raise @parser.error(e, @usage)
    end

    def get_endianness(packet)
      params = @parser.parameters
      max_options = @usage.count("<")
      if params[max_options - 1]
        endianness = params[max_options - 1].to_s.upcase.intern
        if endianness != :BIG_ENDIAN and endianness != :LITTLE_ENDIAN
          raise @parser.error("Invalid endianness #{endianness}. Must be BIG_ENDIAN or LITTLE_ENDIAN.", @usage)
        end
      else
        endianness = packet.default_endianness
      end
      endianness
    end

    def get_range
      return nil if @parser.keyword.include?('ARRAY')

      data_type = get_data_type()
      return nil unless data_type == :INT or data_type == :UINT or data_type == :FLOAT

      index = append? ? 3 : 4
      return nil if @parser.parameters[index] == 'nil'
      min = ConfigParser.handle_defined_constants(
        @parser.parameters[index].convert_to_value, get_data_type(), get_bit_size()
      )
      max = ConfigParser.handle_defined_constants(
        @parser.parameters[index + 1].convert_to_value, get_data_type(), get_bit_size()
      )
      min..max
    end

    def get_cmd_or_tlm
      index = append? ? 2 : 3
      cmd_or_tlm = @parser.parameters[index].to_s.upcase.intern
      raise ArgumentError, "Unknown type: #{cmd_or_tlm}" unless %i(CMD TLM COMMAND TELEMETRY).include?(cmd_or_tlm)
      cmd_or_tlm
    end

    def get_target_name
      index = append? ? 3 : 4
      @parser.parameters[index].to_s.upcase
    end

    def get_packet_name
      index = append? ? 4 : 5
      @parser.parameters[index].to_s.upcase
    end

    def lookup_packet(cmd_or_tlm, target_name, packet_name)
      if cmd_or_tlm == :CMD or cmd_or_tlm == :COMMAND
        return @packet_config.commands[target_name][packet_name]
      else
        return @packet_config.telemetry[target_name][packet_name]
      end
    end

    def convert_string_value(index)
      # If the default value is 0x<data> (no quotes), it is treated as
      # binary data.  Otherwise, the default value is considered to be a string.
      if @parser.parameters[index].upcase.start_with?("0X") and
        !@parser.line.include?("\"#{@parser.parameters[index]}\"") and
        !@parser.line.include?("\'#{@parser.parameters[index]}\'")
        return @parser.parameters[index].hex_to_byte_string
      else
        return @parser.parameters[index]
      end
    end

    def get_default
      return [] if @parser.keyword.include?('ARRAY')

      index = append? ? 3 : 4
      data_type = get_data_type()
      if data_type == :BOOL
        value = @parser.parameters[index].to_s.upcase
        if value == "TRUE" or value == "FALSE"
          return ConfigParser.handle_true_false(@parser.parameters[index])
        else
          raise @parser.error("Default for BOOL data type must be TRUE or FALSE")
        end
      end
      if data_type == :ARRAY
        value = @parser.parameters[index].to_s
        begin
          value = JSON.parse(value, allow_nan: true)
        rescue Exception
          raise @parser.error("Unparsable value for ARRAY: #{value}")
        end
        if Array === value
          return value
        else
          raise @parser.error("Default for ARRAY data type must be an Array")
        end
      end
      if data_type == :OBJECT
        value = @parser.parameters[index].to_s
        begin
          value = JSON.parse(value, allow_nan: true)
        rescue Exception
          raise @parser.error("Unparsable value for OBJECT: #{value}")
        end
        if Hash === value
          return value
        else
          raise @parser.error("Default for OBJECT data type must be a Hash")
        end
      end
      if data_type == :ANY
        value = @parser.parameters[index].to_s
        if value.length > 0
          begin
            return JSON.parse(value, allow_nan: true)
          rescue Exception
            return value
          end
        else
          return ""
        end
      end
      if data_type == :STRING or data_type == :BLOCK
        return convert_string_value(index)
      else
        if data_type != :DERIVED
          return ConfigParser.handle_defined_constants(
            @parser.parameters[index + 2].convert_to_value, data_type, get_bit_size()
          )
        else
          return @parser.parameters[index + 2].convert_to_value
        end
      end
    end

    def get_id_value(item)
      return nil unless @parser.keyword.include?('ID_')
      data_type = get_data_type
      if data_type == :DERIVED
        raise @parser.error("DERIVED data type not allowed for Identifier")
      end
      # For PARAMETERS the default value is the ID value
      if @parser.keyword.include?("PARAMETER")
        return item.default
      end

      index = append? ? 3 : 4
      if data_type == :BOOL
        value = @parser.parameters[index].to_s.upcase
        if value == "TRUE" or value == "FALSE"
          return ConfigParser.handle_true_false(@parser.parameters[index])
        else
          raise @parser.error("ID Value for BOOL data type must be TRUE or FALSE")
        end
      end
      if data_type == :ARRAY
        value = @parser.parameters[index].to_s
        begin
          value = JSON.parse(value, allow_nan: true)
        rescue Exception
          raise @parser.error("Unparsable value for ARRAY: #{value}")
        end
        if Array === value
          return value
        else
          raise @parser.error("ID Value for ARRAY data type must be an Array")
        end
      end
      if data_type == :OBJECT
        value = @parser.parameters[index].to_s
        begin
          value = JSON.parse(value, allow_nan: true)
        rescue Exception
          raise @parser.error("Unparsable value for OBJECT: #{value}")
        end
        if Hash === value
          return value
        else
          raise @parser.error("ID Value for OBJECT data type must be a Hash")
        end
      end
      if data_type == :ANY
        value = @parser.parameters[index].to_s
        if value.length > 0
          begin
            return JSON.parse(value, allow_nan: true)
          rescue Exception
            return value
          end
        else
          return ""
        end
      end
      if data_type == :STRING or data_type == :BLOCK
        return convert_string_value(index)
      else
        return ConfigParser.handle_defined_constants(
          @parser.parameters[index].convert_to_value, data_type, get_bit_size()
        )
      end
    end

    def get_description
      max_options = @usage.count("<")
      @parser.parameters[max_options - 2] if @parser.parameters[max_options - 2]
    end

    # There are many different usages of the ITEM and PARAMETER keywords so
    # parse the keyword and parameters to generate the correct usage information.
    def get_usage
      usage = "#{@parser.keyword} <ITEM NAME> "
      usage << "<BIT OFFSET> " unless @parser.keyword.include?("APPEND")
      usage << bit_size_usage()
      if not @parser.keyword.include?("STRUCTURE")
        usage << type_usage()
        usage << "<TOTAL ARRAY BIT SIZE> " if @parser.keyword.include?("ARRAY")
        usage << id_usage()
        usage << "<DESCRIPTION (Optional)> <ENDIANNESS (Optional)>"
      else
        usage << "<CMD or TLM> <Target Name> <Packet Name>"
      end
      usage
    end

    def bit_size_usage
      if @parser.keyword.include?("ARRAY")
        "<ARRAY ITEM BIT SIZE> "
      else
        "<BIT SIZE> "
      end
    end

    def type_usage
      keyword = @parser.keyword
      # Item type usage is simple so just return it
      return "<TYPE: INT/UINT/FLOAT/STRING/BLOCK/DERIVED/BOOL/ARRAY/OBJECT/ANY> " if keyword.include?("ITEM")

      # Build up the parameter type usage based on the keyword
      usage = "<TYPE: "
      # ARRAY types don't have min or max or default values
      if keyword.include?("ARRAY")
        usage << "INT/UINT/FLOAT/STRING/BLOCK/BOOL/OBJECT/ANY> "
      else
        begin
          data_type = get_data_type()
        rescue
          # If the data type could not be determined set something
          data_type = :INT
        end
        if data_type == :INT || data_type == :UINT || data_type == :FLOAT || data_type == :DERIVED
          usage << "INT/UINT/FLOAT/DERIVED> <MIN VALUE> <MAX VALUE> "
        else
          usage << "STRING/BLOCK/BOOL/ARRAY/OBJECT/ANY> "
        end
        # ID Values do not have default values
        unless keyword.include?("ID")
          usage << "<DEFAULT_VALUE> "
        end
      end
      usage
    end

    def id_usage
      return '' unless @parser.keyword.include?("ID")

      if @parser.keyword.include?("PARAMETER")
        "<DEFAULT AND ID VALUE> "
      else
        "<ID VALUE> "
      end
    end
  end
end

File: openc3/python/openc3/packets/parsers/packet_item_parser.py
# Copyright 2025 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

from ast import literal_eval

from openc3.config.config_parser import ConfigParser
from openc3.packets.packet_item import PacketItem
from openc3.utilities.extract import convert_to_value, hex_to_byte_string
from openc3.utilities.logger import Logger


class PacketItemParser:
    COMMAND = "Command"
    TELEMETRY = "Telemetry"

    # Parses a packet item definition and creates a new PacketItem

    # This number is a little arbitrary but there are definitely issues at
    # 1 million and you really shouldn't be doing anything this big anyway
    BIG_ARRAY_SIZE = 100_000

    # self.param parser [ConfigParser] Configuration parser
    # self.param packet [Packet] The packet the item should be added to
    # self.param cmd_or_tlm [String] Whether this is a command or telemetry packet
    # self.param warnings [Array<String>] Array of warning strings from PacketConfig
    @classmethod
    def parse(cls, parser, packet_config, packet, cmd_or_tlm, warnings):
        parser = PacketItemParser(parser, packet_config, warnings)
        parser.verify_parameters(cmd_or_tlm)
        return parser.create_packet_item(packet, cmd_or_tlm)

    # self.param parser [ConfigParser] Configuration parser
    # self.param warnings [Array<String>] Array of warning strings from PacketConfig
    def __init__(self, parser, packet_config, warnings):
        self.parser = parser
        self.packet_config = packet_config
        self.warnings = warnings
        self.usage = self._get_usage()

    def verify_parameters(self, cmd_or_tlm):
        if "ITEM" in self.parser.keyword and cmd_or_tlm == PacketItemParser.COMMAND:
            raise self.parser.error("ITEM types are only valid with TELEMETRY", self.usage)
        elif "PARAMETER" in self.parser.keyword and cmd_or_tlm == PacketItemParser.TELEMETRY:
            raise self.parser.error("PARAMETER types are only valid with COMMAND", self.usage)

        # The usage is formatted with brackets <XXX> around each option so
        # count the number of open brackets to determine the number of options
        max_options = self.usage.count("<")
        if "STRUCTURE" in self.parser.keyword:
            self.parser.verify_num_parameters(max_options, max_options, self.usage)
        else:
            # The last two options (description and endianness) are optional
            self.parser.verify_num_parameters(max_options - 2, max_options, self.usage)
        self.parser.verify_parameter_naming(1)  # Item name is the 1st parameter

        # ARRAY items cannot have brackets in their name because brackets are used
        # for array indexing in the UI and would cause confusion
        if "ARRAY" in self.parser.keyword:
            item_name = self.parser.parameters[0]
            if "[" in item_name or "]" in item_name:
                raise self.parser.error(f"ARRAY items cannot have brackets in their name: {item_name}", self.usage)

    def create_packet_item(self, packet, cmd_or_tlm):
        try:
            item_name = self.parser.parameters[0].upper()
            if packet.items.get(item_name):
                msg = f"{packet.target_name} {packet.packet_name} {item_name} redefined."
                Logger.warn(msg)
                self.warnings.append(msg)
            if "STRUCTURE" in self.parser.keyword:
                item = PacketItem(
                    item_name,
                    self._get_bit_offset(),
                    self._get_bit_size(True),
                    "BLOCK",
                    "BIG_ENDIAN",
                    None,
                    "ERROR",  # overflow
                )
            else:
                item = PacketItem(
                    item_name,
                    self._get_bit_offset(),
                    self._get_bit_size(),
                    self._get_data_type(),
                    self._get_endianness(packet),
                    self._get_array_size(),
                    "ERROR",  # overflow
                )
                if cmd_or_tlm == PacketItemParser.COMMAND:
                    item.minimum = self._get_minimum()
                    item.maximum = self._get_maximum()
                    item.default = self._get_default()
                item.id_value = self._get_id_value(item)
                item.description = self._get_description()
            if self._append():
                item = packet.append(item)
            else:
                item = packet.define(item)
            if "STRUCTURE" in self.parser.keyword:
                structure = self._lookup_packet(
                    self._get_cmd_or_tlm(), self._get_target_name(), self._get_packet_name()
                )
                packet.structurize_item(item, structure)
            return item
        except Exception as error:
            raise self.parser.error(error, self.usage)

    def _append(self):
        return "APPEND" in self.parser.keyword

    def _get_data_type(self):
        index = 2 if self._append() else 3
        # _get_data_type is called in usage before we verify
        # the parameters so we must check the length here
        if index >= len(self.parser.parameters):
            return None
        return self.parser.parameters[index].upper()

    def _get_bit_offset(self):
        if self._append():
            return 0
        try:
            return int(self.parser.parameters[1], 0)
        except ValueError as error:
            raise self.parser.error(error, self.usage)

    def _get_bit_size(self, check_structure=False):
        index = 1 if self._append() else 2
        try:
            bit_size = self.parser.parameters[index]
            if not check_structure or str(bit_size).upper() != "DEFINED":
                return int(bit_size, 0)
            else:
                structure = self._lookup_packet(
                    self._get_cmd_or_tlm(), self._get_target_name(), self._get_packet_name()
                )
                return structure.defined_length_bits

        except ValueError as error:
            raise self.parser.error(error, self.usage)

    def _get_array_size(self):
        if "ARRAY" not in self.parser.keyword:
            return None

        try:
            index = 3 if self._append() else 4
            array_bit_size = int(self.parser.parameters[index], 0)
            items = int(array_bit_size / self._get_bit_size())
            if items >= PacketItemParser.BIG_ARRAY_SIZE:
                warning = (
                    "Performance Issue!\n"
                    f"In {self.parser.filename}:{self.parser.line_number} your definition of=\n"
                    f"{self.parser.line}\n"
                    f"creates an array with {items} elements. Consider creating a BLOCK if this is binary data."
                )
                Logger.warn(warning)
                self.warnings.append(warning)
            return array_bit_size
        except ValueError as error:
            raise self.parser.error(error, self.usage)

    def _get_endianness(self, packet):
        params = self.parser.parameters
        max_options = self.usage.count("<")
        if len(params) >= max_options:
            endianness = params[max_options - 1].upper()
            if endianness != "BIG_ENDIAN" and endianness != "LITTLE_ENDIAN":
                raise self.parser.error(
                    f"Invalid endianness {endianness}. Must be BIG_ENDIAN or LITTLE_ENDIAN.",
                    self.usage,
                )
        else:
            endianness = packet.default_endianness
        return endianness

    def _get_minimum(self):
        if "ARRAY" in self.parser.keyword:
            return None

        data_type = self._get_data_type()
        if data_type != "INT" and data_type != "UINT" and data_type != "FLOAT":
            return None

        index = 3 if self._append() else 4
        value = self.parser.parameters[index]
        if str(value).upper() == "NIL" or str(value).upper() == "NONE":
            return None
        return ConfigParser.handle_defined_constants(
            convert_to_value(value),
            self._get_data_type(),
            self._get_bit_size(),
        )

    def _get_maximum(self):
        if "ARRAY" in self.parser.keyword:
            return None

        data_type = self._get_data_type()
        if data_type != "INT" and data_type != "UINT" and data_type != "FLOAT":
            return None

        index = 4 if self._append() else 5
        value = self.parser.parameters[index]
        if str(value).upper() == "NIL" or str(value).upper() == "NONE":
            return None
        return ConfigParser.handle_defined_constants(
            convert_to_value(value),
            self._get_data_type(),
            self._get_bit_size(),
        )

    def _get_cmd_or_tlm(self):
        index = 2 if self._append() else 3
        cmd_or_tlm = str(self.parser.parameters[index])
        if cmd_or_tlm not in ["CMD", "TLM", "COMMAND", "TELEMETRY"]:
            raise TypeError(f"Unknown type: {cmd_or_tlm}")
        return cmd_or_tlm

    def _get_target_name(self):
        index = 3 if self._append() else 4
        return str(self.parser.parameters[index]).upper()

    def _get_packet_name(self):
        index = 4 if self._append() else 5
        return str(self.parser.parameters[index]).upper()

    def _lookup_packet(self, cmd_or_tlm, target_name, packet_name):
        if cmd_or_tlm == "CMD" or cmd_or_tlm == "COMMAND":
            return self.packet_config.commands[target_name][packet_name]
        else:
            return self.packet_config.telemetry[target_name][packet_name]

    def _convert_string_value(self, index):
        # If the default value is 0x<data> (no quotes), it is treated as
        # binary data. Otherwise, the default value is considered to be a string.
        if (
            self.parser.parameters[index].upper().startswith("0X")
            and f'"{self.parser.parameters[index]}"' not in self.parser.line
            and f"'{self.parser.parameters[index]}'" not in self.parser.line
        ):
            return hex_to_byte_string(self.parser.parameters[index])
        else:
            return self.parser.parameters[index]

    def _get_default(self):
        if "ARRAY" in self.parser.keyword:
            return []

        index = 3 if self._append() else 4
        data_type = self._get_data_type()
        if data_type == "BOOL":
            value = str(self.parser.parameters[index]).upper()
            if value == "TRUE" or value == "FALSE":
                return ConfigParser.handle_true_false(self.parser.parameters[index])
            else:
                raise self.parser.error("Default for BOOL data type must be TRUE or FALSE")
        if data_type == "ARRAY":
            value = str(self.parser.parameters[index])
            try:
                value = literal_eval(value)
            except Exception:
                raise self.parser.error(f"Unparsable value for ARRAY: {value}")
            if isinstance(value, list):
                return value
            else:
                raise self.parser.error("Default for ARRAY data type must be an Array")
        if data_type == "OBJECT":
            value = str(self.parser.parameters[index])
            try:
                value = literal_eval(value)
            except Exception:
                raise self.parser.error(f"Unparsable value for OBJECT: {value}")
            if isinstance(value, dict):
                return value
            else:
                raise self.parser.error("Default for OBJECT data type must be a Hash")
        if data_type == "ANY":
            value = str(self.parser.parameters[index])
            if len(value) > 0:
                try:
                    return literal_eval(value)
                except Exception:
                    return value
            else:
                return ""
        if data_type == "STRING" or data_type == "BLOCK":
            return self._convert_string_value(index)
        else:
            if data_type != "DERIVED":
                return ConfigParser.handle_defined_constants(
                    convert_to_value(self.parser.parameters[index + 2]),
                    self._get_data_type(),
                    self._get_bit_size(),
                )
            else:
                return convert_to_value(self.parser.parameters[index + 2])

    def _get_id_value(self, item):
        if "ID_" not in self.parser.keyword:
            return None
        data_type = self._get_data_type()
        if data_type == "DERIVED":
            raise self.parser.error("DERIVED data type not allowed for Identifier")
        # For PARAMETERS the default value is the ID value
        if "PARAMETER" in self.parser.keyword:
            return item.default

        index = 3 if self._append() else 4
        if data_type == "BOOL":
            value = str(self.parser.parameters[index]).upper()
            if value == "TRUE" or value == "FALSE":
                return ConfigParser.handle_true_false(self.parser.parameters[index])
            else:
                raise self.parser.error("ID Value for BOOL data type must be TRUE or FALSE")
        if data_type == "ARRAY":
            value = str(self.parser.parameters[index])
            try:
                value = literal_eval(value)
            except Exception:
                raise self.parser.error(f"Unparsable value for ARRAY: {value}")
            if isinstance(value, list):
                return value
            else:
                raise self.parser.error("ID Value for ARRAY data type must be an Array")
        if data_type == "OBJECT":
            value = str(self.parser.parameters[index])
            try:
                value = literal_eval(value)
            except Exception:
                raise self.parser.error(f"Unparsable value for OBJECT: {value}")
            if isinstance(value, dict):
                return value
            else:
                raise self.parser.error("ID Value for OBJECT data type must be a Hash")
        if data_type == "ANY":
            value = str(self.parser.parameters[index])
            if len(value) > 0:
                try:
                    return literal_eval(value)
                except Exception:
                    return value
            else:
                return ""
        if data_type == "STRING" or data_type == "BLOCK":
            return self._convert_string_value(index)
        else:
            return ConfigParser.handle_defined_constants(
                convert_to_value(self.parser.parameters[index]),
                self._get_data_type(),
                self._get_bit_size(),
            )

    def _get_description(self):
        max_options = self.usage.count("<")
        if len(self.parser.parameters) >= max_options - 1:
            return self.parser.parameters[max_options - 2]
        return None

    # There are many different usages of the ITEM and PARAMETER keywords so
    # parse the keyword and parameters to generate the correct usage information.
    def _get_usage(self):
        keyword = self.parser.keyword or ""
        usage = f"{keyword} <ITEM NAME> "
        if "APPEND" not in keyword:
            usage += "<BIT OFFSET> "
        usage += self._bit_size_usage()
        if "STRUCTURE" not in self.parser.keyword:
            usage += self._type_usage()
            if "ARRAY" in keyword:
                usage += "<TOTAL ARRAY BIT SIZE> "
            usage += self._id_usage()
            usage += "<DESCRIPTION (Optional)> <ENDIANNESS (Optional)>"
        else:
            usage += "<CMD or TLM> <Target Name> <Packet Name>"
        return usage

    def _bit_size_usage(self):
        keyword = self.parser.keyword or ""
        if "ARRAY" in keyword:
            return "<ARRAY ITEM BIT SIZE> "
        else:
            return "<BIT SIZE> "

    def _type_usage(self):
        keyword = self.parser.keyword or ""
        # Item type usage is simple so just return it
        if "ITEM" in keyword:
            return "<TYPE: INT/UINT/FLOAT/STRING/BLOCK/DERIVED/BOOL/ARRAY/OBJECT/ANY> "

        # Build up the parameter type usage based on the keyword
        usage = "<TYPE: "
        # ARRAY types don't have min or max or default values
        if "ARRAY" in keyword:
            usage += "INT/UINT/FLOAT/STRING/BLOCK/BOOL/OBJECT/ANY> "
        else:
            try:
                data_type = self._get_data_type()
            except TypeError:
                # If the data type could not be determined set something
                data_type = "INT"
            if data_type == "INT" or data_type == "UINT" or data_type == "FLOAT" or data_type == "DERIVED":
                usage += "INT/UINT/FLOAT/DERIVED> <MIN VALUE> <MAX VALUE> "
            else:
                usage += "STRING/BLOCK/BOOL/ARRAY/OBJECT/ANY> "
            # ID Values do not have default values
            if "ID" not in keyword:
                usage += "<DEFAULT_VALUE> "
        return usage

    def _id_usage(self):
        keyword = self.parser.keyword or ""
        if "ID" not in keyword:
            return ""

        if "PARAMETER" in keyword:
            return "<DEFAULT AND ID VALUE> "
        else:
            return "<ID VALUE> "
